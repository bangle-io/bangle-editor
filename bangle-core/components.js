import { coreMarkPlugins, coreMarkSpec } from './marks/index';
import { coreNodeSpec, coreNodePlugins } from './nodes/index';
import * as history from './extensions/history';

export * from './nodes/index';
export * from './marks/index';
export * as history from './extensions/history';

export function coreSpec(opts = {}) {
  return [
    ...coreMarkSpec(opts.node),
    ...coreNodeSpec(opts.node),
    history.spec(opts.history),
  ];
}

export function corePlugins(opts = {}) {
  return [
    ...coreMarkPlugins(opts.node),
    ...coreNodePlugins(opts.node),
    history.plugins(opts.history),
  ];
}
