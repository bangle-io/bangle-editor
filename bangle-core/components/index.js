// marks
import * as bold from './bold';
import * as code from './code';
import * as italic from './italic';
import * as link from './link';
import * as strike from './strike';
import * as underline from './underline';

// nodes
import * as doc from './doc';
import * as paragraph from './paragraph';
import * as text from './text';
import * as blockquote from './blockquote';
import * as bulletList from './bullet-list';
import * as codeBlock from './code-block';
import * as hardBreak from './hard-break';
import * as heading from './heading';
import * as horizontalRule from './horizontal-rule';
import * as listItem from './list-item/list-item';
import * as orderedList from './ordered-list';
import * as todoItem from './todo-item';
import * as todoList from './todo-list';
import * as image from './image';

// components
import * as history from './history';

// marks
export * as bold from './bold';
export * as code from './code';
export * as italic from './italic';
export * as link from './link';
export * as strike from './strike';
export * as underline from './underline';

// nodes
export * as doc from './doc';
export * as paragraph from './paragraph';
export * as text from './text';
export * as blockquote from './blockquote';
export * as bulletList from './bullet-list';
export * as codeBlock from './code-block';
export * as hardBreak from './hard-break';
export * as heading from './heading';
export * as horizontalRule from './horizontal-rule';
export * as listItem from './list-item/list-item';
export * as orderedList from './ordered-list';
export * as todoItem from './todo-item';
export * as todoList from './todo-list';
export * as image from './image';

// components
export * as history from './history';

export function coreSpec(opts = {}) {
  return [...coreMarkSpec(opts), ...coreNodeSpec(opts)];
}

export function corePlugins(opts = {}) {
  return [
    ...coreMarkPlugins(opts),
    ...coreNodePlugins(opts),
    history.plugins(opts.history),
  ];
}

function coreMarkSpec(options = {}) {
  return [
    bold.spec(options.bold),
    code.spec(options.code),
    italic.spec(options.italic),
    strike.spec(options.strike),
    link.spec(options.link),
    underline.spec(options.underline),
  ];
}

function coreMarkPlugins(options = {}) {
  return [
    bold.plugins(options.bold),
    code.plugins(options.code),
    italic.plugins(options.italic),
    strike.plugins(options.strike),
    link.plugins(options.link),
    underline.plugins(options.underline),
  ];
}

function coreNodeSpec(options = {}) {
  return [
    doc.spec(options.doc),
    text.spec(options.text),
    paragraph.spec(options.paragraph),
    blockquote.spec(options.blockquote),
    bulletList.spec(options.bulletList),
    codeBlock.spec(options.codeBlock),
    hardBreak.spec(options.hardBreak),
    heading.spec(options.heading),
    horizontalRule.spec(options.horizontalRule),
    listItem.spec(options.listItem),
    orderedList.spec(options.orderedList),
    todoItem.spec(options.todoItem),
    todoList.spec(options.todoList),
    image.spec(options.image),
  ];
}

function coreNodePlugins(options = {}) {
  return [
    paragraph.plugins(options.paragraph),
    blockquote.plugins(options.blockquote),
    bulletList.plugins(options.bulletList),
    codeBlock.plugins(options.codeBlock),
    hardBreak.plugins(options.hardBreak),
    heading.plugins(options.heading),
    horizontalRule.plugins(options.horizontalRule),
    listItem.plugins(options.listItem),
    orderedList.plugins(options.orderedList),
    todoItem.plugins(options.todoItem),
    todoList.plugins(options.todoList),
    image.plugins(options.image),
  ];
}
