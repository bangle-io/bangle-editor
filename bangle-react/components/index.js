import * as dinos from './dinos/index';

export * as dinos from './dinos/index';

export const spec = (opts = {}) => {
  return [dinos.spec(opts.dinos)];
};

export const plugins = (opts = {}) => {
  return [dinos.plugins(opts.dinos)];
};

export const commands = {
  dinos: dinos.commands,
};
