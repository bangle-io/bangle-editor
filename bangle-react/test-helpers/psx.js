import React from 'react';
import { psx as banglePsx } from 'bangle-core/test-helpers/index';

export function psx(name, ...args) {
  if (typeof name !== 'string') {
    return React.createElement(name, ...args);
  }
  if (name.startsWith('jsx_')) {
    return React.createElement(name.split('jsx_').join(''), ...args);
  }
  return banglePsx(name, ...args);
}
