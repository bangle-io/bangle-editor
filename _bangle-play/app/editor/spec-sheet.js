import '@banglejs/core/style.css';
import '@banglejs/tooltip/style.css';

import * as collab from '@banglejs/collab/client/collab-extension';
import { emoji } from '@banglejs/emoji/index';
import '@banglejs/emoji/emoji.css';
import '@banglejs/react-menu/style.css';
import '@banglejs/markdown-front-matter/style.css';
import './extensions-override.css';
import { trailingNode } from '@banglejs/trailing-node';
import { timestamp } from '@banglejs/timestamp';
import { SpecRegistry } from '@banglejs/core/spec-registry';
import stopwatch from '@banglejs/react-stopwatch';
import sticker from '@banglejs/react-sticker';
import { emojiSuggestMenu } from '@banglejs/react-menu';
import { markdownFrontMatter } from '@banglejs/markdown-front-matter';
import {
  bold,
  code,
  italic,
  strike,
  link,
  underline,
  doc,
  text,
  paragraph,
  blockquote,
  bulletList,
  codeBlock,
  hardBreak,
  heading,
  horizontalRule,
  listItem,
  orderedList,
  todoItem,
  todoList,
  image,
} from '@banglejs/core/components/index';

export const specRegistry = new SpecRegistry([
  doc.spec({ content: 'frontMatter? block+' }),
  text.spec(),
  paragraph.spec(),
  blockquote.spec(),
  bulletList.spec(),
  codeBlock.spec(),
  hardBreak.spec(),
  heading.spec(),
  horizontalRule.spec(),
  listItem.spec(),
  orderedList.spec(),
  todoItem.spec(),
  todoList.spec(),
  image.spec(),
  bold.spec(),
  code.spec(),
  italic.spec(),
  strike.spec(),
  link.spec(),
  underline.spec(),
  collab.spec(),
  emoji.spec(),
  emojiSuggestMenu.spec(),
  stopwatch.spec(),
  trailingNode.spec(),
  timestamp.spec(),
  sticker.spec(),
  markdownFrontMatter.spec(),
]);
