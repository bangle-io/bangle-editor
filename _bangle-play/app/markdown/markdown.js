import * as markdown from '@banglejs/markdown/index';
import { defaultMarkdownItTokenizer } from '@banglejs/markdown/index';
import { emojiMarkdownItPlugin } from '@banglejs/emoji/index';
import { specRegistry } from 'bangle-play/app/editor/spec-sheet';
import { frontMatterMarkdownItPlugin } from '@banglejs/markdown-front-matter';

const parser = markdown.markdownParser(
  specRegistry,
  defaultMarkdownItTokenizer
    .use(emojiMarkdownItPlugin)
    .use(frontMatterMarkdownItPlugin),
);

const serializer = markdown.markdownSerializer(specRegistry);

export const markdownParser = (markdownStr) => parser.parse(markdownStr);
export const markdownSerializer = (doc) => serializer.serialize(doc);
