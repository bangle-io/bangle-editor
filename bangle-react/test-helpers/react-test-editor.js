/**
 * @jest-environment jsdom
 */
import React from 'react';
import { TextSelection } from 'prosemirror-state';
import { SpecSheet } from 'bangle-core/spec-sheet';
import { render } from '@testing-library/react';
import { ReactEditor } from 'bangle-react/react-editor';
import { corePlugins } from 'bangle-core/index';
import { getDocLabels } from 'bangle-core/test-helpers/index';

const defaultSpecSheet = new SpecSheet();
const defaultPlugins = corePlugins();

export function reactTestEditor(
  {
    specSheet = defaultSpecSheet,
    plugins = defaultPlugins,
    renderNodeViews,
  } = {},
  testId = 'test-editor',
) {
  if (!(specSheet instanceof SpecSheet)) {
    throw new Error('Need to be specsheet');
  }

  return async (testDoc) => {
    let editor;
    const onReady = (_editor) => {
      editor = _editor;
    };

    const editorProps = {
      attributes: { class: 'bangle-editor content' },
    };

    const _options = {
      id: 'test-editor',
      testId,
      editorProps,
      ...{ specSheet, plugins },
    };

    const result = render(
      <ReactEditor
        options={_options}
        onReady={onReady}
        renderNodeViews={renderNodeViews}
      />,
    );

    await result.findByTestId(testId);

    let view = editor.view;

    let posLabels;

    if (testDoc) {
      posLabels = updateDoc(testDoc);
    }

    function updateDoc(doc) {
      if (!doc) {
        return;
      }
      const editorView = view;
      const dispatch = editorView.dispatch;
      const defaultDoc = doc(editorView.state.schema);
      const tr = editorView.state.tr.replaceWith(
        0,
        editorView.state.doc.nodeSize - 2,
        defaultDoc.content,
      );
      tr.setMeta('addToHistory', false);
      dispatch(tr);

      const positionExists = (position) => typeof position === 'number';
      const posLabels = getDocLabels(defaultDoc);
      if (posLabels) {
        if (positionExists(posLabels['[]'])) {
          setTextSelection(editorView, posLabels['[]']);
        } else if (
          positionExists(posLabels['[']) ||
          positionExists(posLabels[']'])
        ) {
          if (
            !positionExists(posLabels['[']) ||
            !positionExists(posLabels[']'])
          ) {
            throw new Error('`[``]` must come in pair.');
          }

          setTextSelection(editorView, posLabels['['], posLabels[']']);
        }
      }

      return posLabels;
    }

    return {
      editor: editor,
      container: result.container,
      renderResult: result,
      view: view,
      editorState: view.state,
      schema: view.state.schema,
      // TODO deprecetate editorView
      editorView: view,
      selection: view.state.selection,
      posLabels,
      updateDoc,
    };
  };
}

function setTextSelection(view, anchor, head) {
  const { state } = view;
  const tr = state.tr.setSelection(
    TextSelection.create(state.doc, anchor, head),
  );
  view.dispatch(tr);
}
