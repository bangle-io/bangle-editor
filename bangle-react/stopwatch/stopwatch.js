import React from 'react';

import { serializeAtomNodeToMdLink } from 'bangle-plugins/markdown/markdown-serializer';
import { keymap } from 'prosemirror-keymap';
import { NodeView } from 'bangle-core/node-view';
import { domSerializationHelpers } from 'bangle-core/dom-serialization-helpers';

const LOG = false;

function log(...args) {
  if (LOG) {
    console.log('stopwatch/index.js:', ...args);
  }
}

const name = 'stopwatch';

export const spec = specFactory;
export const plugins = pluginsFactory;
export const commands = {};

function specFactory() {
  const spec = {
    name,
    type: 'node',
    schema: {
      attrs: {
        'data-stopwatch': {
          default: {
            startTime: 0,
            stopTime: 0,
          },
        },
        'data-bangle-name': {
          default: name,
        },
      },
      inline: true,
      group: 'inline',
      draggable: true,
      atom: true,
    },
    markdown: {
      toMarkdown: (state, node) => {
        const string = serializeAtomNodeToMdLink(name, node.attrs);
        state.write(string);
      },
    },
  };

  spec.schema = {
    ...spec.schema,
    ...domSerializationHelpers(name, { tag: 'span' }),
  };

  return spec;
}

function pluginsFactory(opts = {}) {
  return [
    keymap({
      'Shift-Ctrl-s': insertStopwatch(),
    }),
    NodeView.createPlugin({ name, containerDOM: ['span'] }),
  ];
}

export class Stopwatch extends React.Component {
  state = {
    counter: 0,
  };

  isPaused = () => {
    const { stopTime } = this.getAttrs();

    return stopTime === 0 || stopTime > 0;
  };
  componentDidMount() {
    this.interval = setInterval(() => {
      log('setting interval');
      if (!this.isPaused()) {
        requestAnimationFrame(() => this.incrementCounter());
      }
    }, 1000);
  }

  componentWillUnmount() {
    if (this.interval) {
      clearInterval(this.interval);
      this.interval = null;
    }
  }

  updateAttrs({ stopTime, startTime }) {
    this.props.updateAttrs({
      'data-stopwatch': {
        stopTime,
        startTime,
      },
    });
  }

  getAttrs() {
    const { stopTime, startTime } = this.props.node.attrs['data-stopwatch'];

    return {
      stopTime,
      startTime,
    };
  }

  incrementCounter = () => {
    this.setState({
      counter: this.state.counter + 1,
    });
  };

  render() {
    const { selected } = this.props;
    const { stopTime, startTime } = this.getAttrs();
    const now = Date.now();

    let endTime = stopTime ? stopTime : now;

    // the initial values
    if (stopTime === 0 && startTime === 0) {
      endTime = 0;
    }

    const isPaused = this.isPaused();

    return (
      <span
        className="stopwatch"
        contentEditable={false}
        style={{
          backgroundColor: isPaused ? 'pink' : '#00CED1',
          outline: selected ? '2px solid blue' : null,
          borderRadius: 10,
          padding: '1px 2px 1px 2px',
          margin: '1px 2px',
          fontWeight: 500,
          fontFamily: 'monospace',
        }}
        onClick={() => {
          if (!isPaused) {
            this.updateAttrs({ stopTime: now, startTime });
            return;
          }

          // resume a stopped stopwatch
          this.updateAttrs({
            startTime: startTime + (now - stopTime),
            stopTime: null,
          });
        }}
      >
        ‚è≤{formatTime(((endTime - startTime) / 1000).toFixed(0))}
      </span>
    );
  }
}

function formatTime(secs) {
  var sec_num = parseInt(secs, 10);
  var hours = Math.floor(sec_num / 3600) % 24;
  var minutes = Math.floor(sec_num / 60) % 60;
  var days = Math.floor(sec_num / (24 * 3600));
  var seconds = sec_num % 60;
  const result = [hours, minutes, seconds]
    .map((v) => (v < 10 ? '0' + v : v))
    .join(':');

  return days > 0 ? days + 'd ' + result : result;
}

export function insertStopwatch() {
  return function (state, dispatch) {
    let stopwatchType = state.schema.nodes[name];
    let { $from } = state.selection,
      index = $from.index();
    if (!$from.parent.canReplaceWith(index, index, stopwatchType)) {
      return false;
    }
    if (dispatch) {
      dispatch(state.tr.replaceSelectionWith(stopwatchType.create()));
    }
    return true;
  };
}
