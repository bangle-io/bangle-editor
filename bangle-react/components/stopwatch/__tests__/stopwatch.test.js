/**
 * @jest-environment jsdom
 */
/** @jsx psx */
import { fireEvent, wait } from '@testing-library/react';
import { psx } from 'bangle-react/test-helpers/psx';
import { sendKeyToPm } from 'bangle-core/test-helpers/index';
import { markdownSerializer } from 'bangle-plugins/markdown/markdown-serializer';
import { corePlugins, coreSpec } from 'bangle-core/components/index';
import { SpecSheet } from 'bangle-core/spec-sheet';
import { stopwatch } from '../index';
import { Stopwatch } from '../stopwatch';
import { reactTestEditor } from 'bangle-react/test-helpers/react-test-editor';

const specSheet = new SpecSheet([...coreSpec(), stopwatch.spec({})]);
const plugins = [...corePlugins(), stopwatch.plugins({})];

const renderNodeViews = jest.fn(({ node, ...args }) => {
  if (node.type.name === 'stopwatch') {
    return <Stopwatch node={node} {...args} />;
  }
  throw new Error('Unknown node');
});

const testEditor = reactTestEditor({
  specSheet,
  plugins,
  renderNodeViews,
});

const dateNow = Date.now;
beforeEach(() => {
  Date.now = dateNow;
});

test('Keyboard shortcut works', async () => {
  Date.now = jest.fn(() => 0);
  const { view } = await testEditor(
    <doc>
      <ul>
        <li>
          <para>foo[]bar</para>
        </li>
      </ul>
    </doc>,
  );

  sendKeyToPm(view, 'Shift-Ctrl-s');

  expect(view.state).toEqualDocAndSelection(
    <doc>
      <ul>
        <li>
          <para>
            foo
            <stopwatch data-stopwatch-time="0" />
            bar
          </para>
        </li>
      </ul>
    </doc>,
  );
});

test('Renders react component correctly', async () => {
  Date.now = jest.fn(() => 0);
  const { container, view } = await testEditor(
    <doc>
      <ul>
        <li>
          <para>foo[]bar</para>
        </li>
      </ul>
    </doc>,
  );
  sendKeyToPm(view, 'Shift-Ctrl-s');
  expect(container.querySelector(`.stopwatch`)).toMatchSnapshot();
});

test('Renders clicking correctly', async () => {
  Date.now = jest.fn(() => 1);
  const { view, renderResult } = await testEditor(
    <doc>
      <ul>
        <li>
          <para>
            foo
            <stopwatch data-stopwatch-time="10" />
            bar
          </para>
        </li>
      </ul>
    </doc>,
  );

  let item = await renderResult.findByText(/⏲/);

  expect(view.state.doc.nodeAt(6).attrs).toEqual({
    'data-stopwatch': {
      startTime: 0,
      stopTime: 0,
    },
    'data-type': 'stopwatch',
  });

  Date.now = jest.fn(() => 150);
  fireEvent.click(item);

  expect(view.state.doc.nodeAt(6).attrs).toEqual({
    'data-stopwatch': {
      startTime: 1,
      stopTime: null,
    },
    'data-type': 'stopwatch',
  });
});

describe('markdown', () => {
  const serialize = async (doc) => {
    let content = doc;
    if (typeof doc === 'function') {
      content = doc(specSheet.schema);
    }
    return markdownSerializer(specSheet).serialize(content);
  };

  test('markdown serialization', async () => {
    const md = await serialize(
      <doc>
        <para>
          hello world
          <stopwatch data-stopwatch-time="10" />
        </para>
      </doc>,
    );
    expect(md).toMatchInlineSnapshot(
      `"hello world[$stopwatch](bangle://data-stopwatch=%7B%22startTime%22%3A0%2C%22stopTime%22%3A0%7D&data-type=%22stopwatch%22)"`,
    );
  });
});
