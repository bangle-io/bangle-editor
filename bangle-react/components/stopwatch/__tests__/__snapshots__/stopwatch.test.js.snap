// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders react component correctly 1`] = `
<span
  class="stopwatch"
  contenteditable="false"
  style="background-color: pink; border-radius: 10px; padding: 1px 2px 1px 2px; margin: 1px 2px; font-weight: 500; font-family: monospace;"
>
  ‚è≤
  00:00:00
</span>
`;
